include/seqan3/io/stream/tokenisation.hpp

```diff
@@ -0,0 +1,98 @@
+// modification, are permitted provided that the following conditions are met:
+//
+//     * Redistributions of source code must retain the above copyright
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ============================================================================
+
+/*!\file
+ * \brief Provides tokenisation functionality.
+ * \author Rene Rahn <rene.rahn AT fu-berlin.de>
+ */
+
+#pragma once
+
+#include <functional>
+
+#include <seqan3/core/concept/core.hpp>
+#include <seqan3/core/concept/iterator.hpp>
+#include <seqan3/core/metafunction/range.hpp>
+#include <seqan3/io/exception.hpp>
+#include <seqan3/range/concept.hpp>
+
+namespace seqan3::detail
+{
+// ----------------------------------------------------------------------------
+// transfer_data
+// ----------------------------------------------------------------------------
+
+//!\cond
+//TODO(rrahn): Replace by anonymous lambda expression once https://gcc.gnu.org/bugzilla/show_bug.cgi?id=85513 is fixed.
+auto invocable_dummy = [](auto){};
+//!\endcond
+
+template <typename            out_iterator_type,
+          input_range_concept input_rng_type,
+          typename            delimiter_type,
+          typename            asserter_type  = decltype(invocable_dummy)>
+//!\cond
+    requires output_iterator_concept<std::remove_reference_t<out_iterator_type>, char> &&
+             predicate_concept<std::remove_reference_t<delimiter_type>, char> &&
+             invocable_concept<std::remove_reference_t<asserter_type>, char>
+//!\endcond
+void transfer_data(out_iterator_type && receiver,
+                   input_rng_type    && transmitter,
+                   delimiter_type    && delim,
```

Comment:
I would prefer the name "source" and "target", or "input_range" and "output_it" and also this order, as it is more intuitive (from source to target).

Also if the asserter is called asserter, maybe the delimiter should instead be called "aborter" or "terminator"? "delim" is very general, especially in tokenisation I would expect to have multiple delimiters in one range, but here it is clearly meant as a terminator, or not?

Reply:
> Obviously, you parse the input range by the delimiters, right?

That depends. If you have a tab-separated file, of course there is a delimiter and the terminator of of each parse call (other than the last) would be the delimiter. But this doesn't have to be the case, e.g what's the delimiter of FastQ files?

Or to be more precise: a "delimiter" is something that separates two sides, it has a relation to both. A "terminator" on the other hand has only a relation to the sequence left of it. So in our case "terminator" would be more concise IMHO.